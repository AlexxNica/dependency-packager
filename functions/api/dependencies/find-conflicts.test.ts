import { ILambdaResponse } from "../";
import findConflicts from "./find-conflicts";

describe("find-conflicts", () => {
  function buildDependencies(
    name: string,
    version: string,
    dependencies: {
      [name: string]: {
        semver: string;
        resolved: string;
        parents: string[];
        entries: string[];
      };
    },
  ): ILambdaResponse {
    return {
      aliases: {},
      contents: {},
      dependency: {
        name,
        version,
      },
      peerDependencies: {},
      dependencyDependencies: dependencies,
    };
  }

  it("combines non-conflicting dependencies", () => {
    const dependencies = [
      buildDependencies("test1", "0.0.0", {
        "react-dom": {
          semver: "^1.0.0",
          resolved: "1.0.0",
          parents: ["test1"],
          entries: [],
        },
        react: {
          semver: "^2.0.0",
          resolved: "2.0.0",
          parents: ["test1"],
          entries: [],
        },
      }),
      buildDependencies("test2", "0.0.0", {
        "react-dom": {
          semver: "^1.0.0",
          resolved: "1.0.0",
          parents: ["test2"],
          entries: [],
        },
      }),
    ];

    expect(findConflicts(dependencies)).toMatchSnapshot();
  });

  it("creates separate entries for conflicting versions with different entrypoints", () => {
    const dependencies = [
      buildDependencies("test1", "0.0.0", {
        "react-dom": {
          semver: "^1.0.0",
          resolved: "1.0.0",
          parents: ["test1"],
          entries: ["a"],
        },
        react: {
          semver: "^2.0.0",
          resolved: "2.0.0",
          parents: ["test1"],
          entries: [],
        },
      }),
      buildDependencies("test2", "0.0.0", {
        "react-dom": {
          semver: "^1.0.0",
          resolved: "1.3.5",
          parents: ["test2"],
          entries: ["b"],
        },
      }),
    ];

    expect(findConflicts(dependencies)).toMatchSnapshot();
  });

  it("doesn't create separate entries for semver matching when the entry points match", () => {
    const dependencies = [
      buildDependencies("test1", "0.0.0", {
        "react-dom": {
          semver: "^1.0.0",
          resolved: "1.0.0",
          parents: ["test1"],
          entries: ["a"],
        },
        react: {
          semver: "^2.0.0",
          resolved: "2.0.0",
          parents: ["test1"],
          entries: [],
        },
      }),
      buildDependencies("test2", "0.0.0", {
        "react-dom": {
          semver: "^1.0.0",
          resolved: "1.3.5",
          parents: ["test2"],
          entries: ["a"],
        },
      }),
    ];

    expect(findConflicts(dependencies)).toMatchSnapshot();
  });
});
