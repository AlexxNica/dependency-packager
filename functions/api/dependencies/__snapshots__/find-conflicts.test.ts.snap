// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`find-conflicts combines non-conflicting dependencies 1`] = `
Object {
  "react": Object {
    "2.0.0": Object {
      "entries": Array [],
      "parentRootDependencies": Array [
        "test1",
      ],
      "parents": Array [
        "test1",
      ],
      "semver": "^2.0.0",
    },
  },
  "react-dom": Object {
    "1.0.0": Object {
      "entries": Array [],
      "parentRootDependencies": Array [
        "test1",
        "test2",
      ],
      "parents": Array [
        "test1",
        "test2",
      ],
      "semver": "^1.0.0",
    },
  },
}
`;

exports[`find-conflicts creates separate entries for conflicting versions with different entrypoints 1`] = `
Object {
  "react": Object {
    "2.0.0": Object {
      "entries": Array [],
      "parentRootDependencies": Array [
        "test1",
      ],
      "parents": Array [
        "test1",
      ],
      "semver": "^2.0.0",
    },
  },
  "react-dom": Object {
    "1.0.0": Object {
      "entries": Array [
        "a",
      ],
      "parentRootDependencies": Array [
        "test1",
      ],
      "parents": Array [
        "test1",
      ],
      "semver": "^1.0.0",
    },
    "1.3.5": Object {
      "entries": Array [
        "b",
      ],
      "parentRootDependencies": Array [
        "test2",
      ],
      "parents": Array [
        "test2",
      ],
      "semver": "^1.0.0",
    },
  },
}
`;

exports[`find-conflicts doesn't create separate entries for semver matching when the entry points match 1`] = `
Object {
  "react": Object {
    "2.0.0": Object {
      "entries": Array [],
      "parentRootDependencies": Array [
        "test1",
      ],
      "parents": Array [
        "test1",
      ],
      "semver": "^2.0.0",
    },
  },
  "react-dom": Object {
    "1.3.5": Object {
      "entries": Array [
        "a",
      ],
      "parentRootDependencies": Array [
        "test1",
        "test2",
      ],
      "parents": Array [
        "test1",
        "test2",
      ],
      "semver": "^1.0.0",
    },
  },
}
`;
